text/3698-declarative-derive-macros.md

```diff
@@ -0,0 +1,144 @@
+- Feature Name: `declarative_derive_macros`
+- Start Date: 2024-09-20
+- RFC PR: [rust-lang/rfcs#3698](https://github.com/rust-lang/rfcs/pull/3698)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Support implementing `derive(Trait)` via a `macro_rules!` macro.
+
+# Motivation
+[motivation]: #motivation
+
+Many crates support deriving their traits with `derive(Trait)`. Today, this
+requires defining proc macros, in a separate crate, typically with several
+additional dependencies adding substantial compilation time, and typically
+guarded by a feature that users need to remember to enable.
+
+However, many common cases of derives don't require any more power than an
+ordinary `macro_rules!` macro. Supporting these common cases would allow many
+crates to avoid defining proc macros, reduce dependencies and compilation time,
+and provide these macros unconditionally without requiring the user to enable a
+feature.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+You can define a macro to implement `derive(MyTrait)` by defining a
+`macro_rules!` macro with the `#[macro_derive]` attribute. Such a macro can
+create new items based on a struct, enum, or union. Note that the macro can
+only append new items; it cannot modify the item it was applied to.
+
+For example:
+
+```rust
+trait Answer { fn answer(&self) -> u32; }
+
+#[macro_derive]
+macro_rules! Answer {
+    // Simplified for this example
+    (struct $n:ident $_:tt) => {
```

Comment:
> To be fair, it's very hard to parse all the variety of the Rust struct syntax using the declarative macro matchers, which makes this applicable in very-very simple cases. 

Looking over https://veykril.github.io/tlborm/decl-macros/building-blocks/parsing.html#struct, it doesn't seem too bad.  Enums are pretty bad though.  In both cases, its made slightly easier by the fact that derives operate in append mode vs attribute macros and regular macro_rules which operate in replace mode.  I did call out this complexity on https://github.com/rust-lang/rfcs/pull/3697#discussion_r1779327250, particularly since attribute macros are most likely to be used on functions

Reply:
Well, that guide doesn't show how you'd parse struct's generics and the `where clause`. The `where` clause is especially problematic. I'll elaborate on the use case I had with parsing generic `impl` blocks in https://github.com/rust-lang/rfcs/pull/3697#discussion_r1779327250 comment thread instead

Reply:
Genetics have a lot of variety (bounds, a mix of lifetimes and types, more complicated generic types, `+`) and it is nearly impossible to handle anything more than the most simple cases. `where` clauses are nearly as bad.

It seems pretty likely that we will eventually wind up with new matchers to better support this, though there is no concrete proposal or implementation. See discussion at https://rust-lang.zulipchat.com/#narrow/stream/404510-wg-macros/topic/Spitballing.3A.20generic-dedicated.20matchers and the other linked Zulip streams (help wanted! Feel free to chime in there if interested)